src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,414 @@
+    ) external {
+        Order memory listing = orders[listingId];
+        if (!listing.isListing || listing.creator == address(0)) {
+            // Is a listing, cancelled, completed or never existed
+            revert InvalidListingId(listingId);
+        }
+
+        _acceptOrder(listingId, listing, quantity, additionalFees, additionalFeeRecievers);
+
+        emit ListingAccepted(listingId, msg.sender, listing.tokenContract, quantity);
+    }
+
+    /**
+     * Sells a token.
+     * @param offerId The ID of the listing.
+     * @param quantity The quantity of tokens to sell.
+     * @param additionalFees The additional fees to pay.
+     * @param additionalFeeRecievers The addresses to send the additional fees to.
+     */
+    function acceptOffer(
+        bytes32 offerId,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) external {
+        Order memory offer = orders[offerId];
+        if (offer.isListing || offer.creator == address(0)) {
+            // Is a listing, cancelled, completed or never existed
+            revert InvalidOfferId(offerId);
+        }
+
+        _acceptOrder(offerId, offer, quantity, additionalFees, additionalFeeRecievers);
+
+        emit OfferAccepted(offerId, msg.sender, offer.tokenContract, quantity);
+    }
+
+    function _acceptOrder(
+        bytes32 orderId,
+        Order memory order,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) internal {
+        if (quantity == 0 || quantity > order.quantity) {
+            revert InvalidQuantity();
+        }
+        if (_isExpired(order)) {
+            revert InvalidExpiry();
+        }
+        if (additionalFees.length != additionalFeeRecievers.length) {
+            revert InvalidAdditionalFees();
+        }
+
+        // Calculate payables
+        uint256 totalCost = order.pricePerToken * quantity;
+        (address royaltyRecipient, uint256 royaltyAmount) =
+            getRoyaltyInfo(order.tokenContract, order.tokenId, totalCost);
+
+        address currencyReceiver = order.isListing ? order.creator : msg.sender;
+        address tokenReceiver = order.isListing ? msg.sender : order.creator;
+
+        if (royaltyAmount > 0) {
+            // Transfer royalties
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, royaltyRecipient, royaltyAmount);
+        }
+
+        // Transfer currency
+        TransferHelper.safeTransferFrom(order.currency, tokenReceiver, currencyReceiver, totalCost - royaltyAmount);
+
+        // Transfer additional fees
+        for (uint256 i; i < additionalFees.length; i++) {
+            if (additionalFeeRecievers[i] == address(0) || additionalFees[i] == 0) {
+                revert InvalidAdditionalFees();
+            }
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, additionalFeeRecievers[i], additionalFees[i]);
```

Comment:
If this an offer, does this pull tokens from the address that's offering, right? but the one that executes this method can populate these fields in whatever way they want, right?

if that's the case, then whoever is executing the order can use this mechanism to drain all funds from the order creator

Reply:
Nice catch. These fees should always be paid by the taker (msg.sender), not the tokenReceiver.